ouroboros-network-framework/test/Test/Ouroboros/Network/Server2.hs

```diff
@@ -2366,6 +2370,112 @@ prop_connection_manager_counters serverAcc (ArbDataFlow dataFlow)
+prop_connection_manager_timeouts_enforced :: Int
+                                          -> ArbDataFlow
+                                          -> AbsBearerInfo
+                                          -> MultiNodeScript Int TestAddr
+                                          -> Property
+prop_connection_manager_timeouts_enforced serverAcc (ArbDataFlow dataFlow)
+                                          absBi (MultiNodeScript events) =
+  let trace = runSimTrace sim
+
+      transitionSignal :: [[(Time, AbstractTransitionTrace SimAddr)]]
+      transitionSignal = splitConns
+                      $ getTraceEvents
+                      $ trace
+
+  in verifyAllTimeouts transitionSignal
+  where
+    verifyAllTimeouts :: [[(Time , AbstractTransitionTrace SimAddr)]]
+                      -> Property
+    verifyAllTimeouts [] = property True
+    verifyAllTimeouts (t:tts) =
+      counterexample ("\nConnection transition trace:\n"
+                     ++ intercalate "\n" (map show t)
+                     )
+      $ verifyTimeouts Map.empty t
+      .&&. verifyAllTimeouts tts
+
+    verifyTimeouts :: Map AbstractState Time
+                   -> [(Time , AbstractTransitionTrace SimAddr)]
+                   -> Property
+    verifyTimeouts stateMap [] =
+      counterexample
+        ("These states didn't timeout:\n"
+        ++ show stateMap
+        )
+      $ Map.null stateMap
+    verifyTimeouts stateMap ((t, TransitionTrace _ tt@(Transition from to)):xs)
+      | Map.member from stateMap =
+        let newStateMap  = Map.delete from stateMap
+            newStateMap' = Map.insert to t newStateMap
+            t'           = stateMap Map.! from
+            valid        = case from of
+              InboundIdleSt _  ->
+               counterexample
+                 (errorMsg tt t' t (tProtocolIdleTimeout simTimeouts))
+               $ diffTime t t' <= tProtocolIdleTimeout simTimeouts
+              OutboundDupSt _  ->
+               counterexample
+                 (errorMsg tt t' t (tOutboundIdleTimeout simTimeouts))
+               $ diffTime t t' <= tOutboundIdleTimeout simTimeouts
+              OutboundIdleSt _ ->
+               counterexample
+                 (errorMsg tt t' t (tOutboundIdleTimeout simTimeouts))
+               $ diffTime t t' <= tOutboundIdleTimeout simTimeouts
+              TerminatingSt    ->
+               counterexample
+                 (errorMsg tt t' t (tTimeWaitTimeout simTimeouts))
+               $ diffTime t t' <= tTimeWaitTimeout simTimeouts
+              _                -> property True
+         in case to of
+           InboundIdleSt _       -> valid .&&. verifyTimeouts newStateMap' xs
+           OutboundDupSt Ticking -> valid .&&. verifyTimeouts newStateMap' xs
+           OutboundIdleSt _      -> valid .&&. verifyTimeouts newStateMap' xs
+           TerminatingSt         -> valid .&&. verifyTimeouts newStateMap' xs
+           _                     -> verifyTimeouts newStateMap xs
```

Comment:
I think this is not right.  If we have a transtion `OutboundDupSt Ticking -> OutboundDupSt Expired` we will not do the check.  In all cases of the `to` state we can check either of the two:
* timeout expired
* the transition was triggered by some other event, sooner than the timeout
These two cases would be covered by:
```
case to of
  _ -> valid .&&. verifyTimeouts newStateMap' xs
```
but this will not work for the `InboundIdleState^\tau -> OutboundState^\tau -> OutboundState` sequence.  The first transition would be fine, but for the second we need the time when we transitioned into `InboundIdleState` and not `OutboundState` (which is not `t` when we check the other transition).

Reply:
I think you are right when you say I need to check timeout expired on the catch-all case, thanks!

I didn't think about that transition, I think it is easy to accommodate for that case, we just have to keep the time from the first transition into InboundIdleState as you said :)

Reply:
Addressed!